### PR TITLE
Fix QuantileConfidence::computeBilateralRank

### PR DESCRIPTION
**Pull request by jschueller**
#3182

## Summary by CodeRabbit

* **Bug Fixes**
  * Improved bilateral confidence interval search and probability evaluation for greater numerical robustness and correctness in edge cases, reducing spurious failures.

* **Tests**
  * Expanded regression and randomized tests for bilateral ranks, adding fixed edge cases and stress checks to validate interval ordering and probability constraints.

### CODE REVIEW

**Comment by mbaudin47** [comment]
Here is a review of the code at 3890b36dd2d572afb83831698f3d4972b09223bb. I focus on the defects from my point of view.

### Inability to Include Zero in the Lower Bound

#### 1. Code Segment

```cpp
UnsignedInteger k1 = 0;
Scalar p1 = binomial.computeCDF(k1);
// ...
while (k1 < size)
{
  // find upper bound of cdf increment location
// ...
```

#### 2. The Problem

The value of $k_1$ is initialized in the statement `UnsignedInteger k1 = 0;`. When the `while` loop begins, it would be logical to test whether this value of $k_1$ is suitable, meaning whether it can lead to a $k_2$ value associated with the required coverage level. However, the algorithm does not test this value of $k_1$: instead, a probability increment is searched for starting from this value of $k_1$. This is an algorithm design issue, because if $k_1 = 0$ is suitable, then the search is unnecessary or harmful. Worse, in the event of a difficulty with the algorithm, the value $k_1 = 0$ can be skipped to move to the next value of $k$ leading to a CDF jump. This may be detrimental if $\alpha$ is close to zero, because most of the mass of the binomial is located near small values of $k$ in this case.

#### 3. The Solution

The solution consists of swapping the sections of the algorithm. First, the `while` loop must start by testing the current value of $k_1$: this corresponds to the code `if (p1 + beta_ > 1.0)`. Then, after testing the value of $k_1$, the next CDF jump can be searched for using the algorithm `// find upper bound of cdf increment location`.

### Inadequate Failure Handling

#### 1. Code Segment

```cpp
UnsignedInteger k1Best = size;
UnsignedInteger k2Best = size;
```

#### 2. The Problem

At the beginning of the algorithm, the values of $k_1$ and $k_2$ are initialized. However, no verification of the search's success is performed. Consequently, if the algorithm fails to find a value for $(k_1, k_2)$, the initial values are returned. The variables `k1Best` and `k2Best` are initialized to `size`. If no pair numerically satisfies the condition $p \ge \beta$ (which can occur in conjunction with premature stopping), the loop ends without ever updating these variables. The function silently returns the array `{size, size}` without throwing an exception. This creates a degenerate and invalid interval for downstream computations.

#### 3. The Solution

The solution consists of using a boolean to detect the success of the algorithm. If the algorithm fails to find a pair $(k_1, k_2)$ satisfying the constraint, an exception must be thrown.

### Numerical Stability of the Exponential Search

#### 1. Code Segment

```cpp
// find upper bound of cdf increment location
UnsignedInteger inc = 1;
UnsignedInteger k1Prev = k1;
while (p1 < p1Prev + epsilon)
{
  k1Prev = k1;
  k1 = std::min(k1 + inc, size);
  p1 = binomial.computeCDF(k1);
  inc *= 2;
}
```

#### 2. The Problem

The exponential leap `inc *= 2` creates an integer overflow vulnerability prior to the execution of the boundary clamp. If the multiplier `inc` grows sufficiently large during a prolonged search, the evaluation of `k1 + inc` within the `std::min(k1 + inc, size)` statement will exceed the maximum capacity of `UnsignedInteger`. This arithmetic overflow forces the value to wrap around before the `std::min` function can clamp it to `size`, destroying the index progression and causing unpredictable algorithmic behavior.

#### 3. The Solution

Implement an arithmetic overflow guard before performing the addition. The assignment must structurally verify that the available capacity of the integer type can safely absorb the exponential increment before attempting the bounds check.

```cpp
k1 = (inc < std::numeric_limits<UnsignedInteger>::max() - k1) ? std::min(k1 + inc, size) : size;
```

File: lib/src/Uncertainty/Algorithm/Simulation/QuantileConfidence.cxx [hunk 1]
```diff
@@ -112,20 +112,27 @@ Indices QuantileConfidence::computeBilateralRank(const UnsignedInteger size) con
...
   UnsignedInteger k1Best = size;
   UnsignedInteger k2Best = size;
-  Scalar p1Prev = 0.0;
+  UnsignedInteger k2Prev = 0;
+  Scalar p1Prev = -1.0;
   UnsignedInteger k1 = 0;
+  Scalar p1 = binomial.computeCDF(k1);
+  Scalar p2Prev = 0.0;
+  // we pass increments of cdf in the scale of beta to quantile function
+  const Scalar epsilon = SpecFunc::ScalarEpsilon * beta_;
```
Comment (mbaudin47) [inline lib/src/Uncertainty/Algorithm/Simulation/QuantileConfidence.cxx]: Computing epsilon based on the value of beta is not relevant here because beta is close to 1 in most cases.

```suggestion
  const Scalar epsilon = SpecFunc::ScalarEpsilon;
```

**Comment by mbaudin47** [comment]
This version seems excellent from my point of view. I will make several tests today et let you know about the results.

File: lib/src/Uncertainty/Algorithm/Simulation/QuantileConfidence.cxx [hunk 2]
```diff
@@ -100,67 +100,44 @@ UnsignedInteger QuantileConfidence::computeUnilateralRank(const UnsignedInteger
...
-  UnsignedInteger k2Best = size;
-  Scalar p1Prev = 0.0;
+  Scalar pBest = 2.0;
+  Indices kBest;
   UnsignedInteger k1 = 0;
   while (k1 < size)
   {
-    // find upper bound of cdf increment location
+    // update CDF at k1, stop when it crosses the bound
     Scalar p1 = binomial.computeCDF(k1);
```
Comment (mbaudin47) [inline lib/src/Uncertainty/Algorithm/Simulation/QuantileConfidence.cxx]:
```suggestion
    const Scalar p1 = binomial.computeCDF(k1);
```

**Review by mbaudin47 — CHANGES_REQUESTED** [review]

File: lib/src/Uncertainty/Algorithm/Simulation/QuantileConfidence.cxx [hunk 3]
```diff
@@ -100,67 +100,44 @@ UnsignedInteger QuantileConfidence::computeUnilateralRank(const UnsignedInteger
...
       pBest = p;
-      k1Best = k1;
-      k2Best = k2;
+      kBest = {k1, k2};
     }
-  } // while k1
-  return {k1Best, k2Best};
+
+    // k1 advances right before the next CDF bump at qB(p2-beta), ensure it stays stricly increasing
+    k1 = std::max(binomial.computeScalarQuantile(p2 - beta_) - 1.0, k1 + 1.0);
```
Comment (mbaudin47) [inline lib/src/Uncertainty/Algorithm/Simulation/QuantileConfidence.cxx:135]: This line is ok because the output of computeScalarQuantile is subtracted the floating point number 1.0. Hence, even if the output of computeScalarQuantile is the UnsignedInteger 0, it is implicitly converted into a floating point number. Therefore, there is no risk of wrap-around here. It seems to me that a small comment may clarify this point in order to make future maintenance easier. Otherwise, this subtle point may be overlooked.

File: lib/src/Uncertainty/Algorithm/Simulation/QuantileConfidence.cxx [hunk 4]
```diff
@@ -100,67 +100,44 @@ UnsignedInteger QuantileConfidence::computeUnilateralRank(const UnsignedInteger
...
+  }
+
+  if (!kBest.getSize())
```
Comment (mbaudin47) [inline lib/src/Uncertainty/Algorithm/Simulation/QuantileConfidence.cxx:138]: This implicitly converts an integer into a boolean. I suggest this:
```suggestion
  if (kBest.getSize()==0)
```

Comment (jschueller) [reply lib/src/Uncertainty/Algorithm/Simulation/QuantileConfidence.cxx:138]: this is very common, imho its ok

File: lib/src/Uncertainty/Algorithm/Simulation/QuantileConfidence.cxx [hunk 5]
```diff
@@ -100,67 +100,44 @@ UnsignedInteger QuantileConfidence::computeUnilateralRank(const UnsignedInteger
...
   if (size < minimumSize)
     throw InvalidArgumentException(HERE) << "Cannot compute bilateral rank as size (" << size << ") is lower than minimum size (" << minimumSize << ")";
 
-  // find the indices of interval ]k1, k2] with smallest probability >=beta according to a Binomial(n=size, alpha)
   const Binomial binomial(size, alpha_);
-  Scalar pBest = SpecFunc::MaxScalar;
-  UnsignedInteger k1Best = size;
-  UnsignedInteger k2Best = size;
-  Scalar p1Prev = 0.0;
+  Scalar pBest = 2.0;
```
Comment (mbaudin47) [inline lib/src/Uncertainty/Algorithm/Simulation/QuantileConfidence.cxx:111]: This initialization is confusing.

```suggestion
  Scalar pBest = SpecFunc::MaxScalar;
```

**Comment by mbaudin47** [comment]
The case where the confidence level is equal to the difference of CDF of the optimal ranks is very special: it might be useful to include it into the unit tests. In one of my algorithms, this generates an infinite loop. I suggest to add it into the unit tests.
```python
"""
Compute the confidence level such that it is exactly
equal to the difference of known ranks.
"""
import openturns as ot

n = int(1.0e6)
k1_ref = 49625
k2_ref = 50533
alpha = 0.05

binomial = ot.Binomial(n, alpha)
beta = binomial.computeCDF(k2_ref) - binomial.computeCDF(k1_ref)

algo = ot.QuantileConfidence(alpha, beta)
k1, k2 = algo.computeBilateralRank(n)
assert k1 == k1_ref
assert k2 == k2_ref
p = algo.computeBilateralCoverage(n, k1, k2)
assert p >= beta
````

File: lib/src/Uncertainty/Algorithm/Simulation/QuantileConfidence.cxx (hunk 5 above)
Comment (jschueller) [reply lib/src/Uncertainty/Algorithm/Simulation/QuantileConfidence.cxx:111]: any value above 2 is ok

Comment (mbaudin47) [reply lib/src/Uncertainty/Algorithm/Simulation/QuantileConfidence.cxx:111]: What is magic about 2?

**Comment by jschueller** [comment]
ok, included

**Comment by mbaudin47** [comment]
@adutfoy I suggest to review this PR

**Review by mbaudin47 — APPROVED** [review]

**Comment by jschueller** [comment]
@regislebrun @sofianehaddad  we would need one more reviewer; this blocks #3176

**Comment by mbaudin47** [comment]
@regislebrun @sofianehaddad @adutfoy Can you please review this PR?

**Comment by mbaudin47** [comment]
@regislebrun @sofianehaddad @adutfoy Can you please review this PR?

**Comment by adutfoy** [comment]
This PR is ok for me!
